[Datastore] Improve support for Dask DataFrames

hayesgb opened the pull request:
Improves support for treating Dask DataFrames as DataItems within mlrun

Hedingber commented:
Hi @hayesgb 
Same here regarding the lint CI, please run `make fmt` and `make lint`
Also note that it looks like this PR includes some of the changes from your other PR (https://github.com/mlrun/mlrun/pull/1140) I guess you tested this PR we Azue and needed the other PR contents for it to work for you. Also it has a change of adding the `_drop_reserved_columns` although you don't use it, this is an old function we already removed, so maybe your base branch wasn't updated. please remove any code changes that are unrelated to the actual PR purpose

yaronha commented on mlrun/datastore/base.py:
```diff
@@ -193,14 +193,15 @@ def reader(*args, **kwargs):
 
         fs = self.get_filesystem()
         if fs:
```
@hayesgb i see that fs (`self.get_filesystem()`) = `fsspec.filesystem("az", **self.get_storage_options())` , so why do we need to pass the storage_options again to dask?

hayesgb replied:
When passing a single parquet file (rather than a directory) but loading with Dask, the below by-passes the loading of storage_options.

yaronha replied:
but the single file option uses the fsspec file handle (`fs.open()`) which is supposed to incorporate the options initialized by the `fsspec.filesystem()` i thought?

hayesgb replied:
It does, but it returns a Pandas DataFrame.  If the code within the consuming Python function expects a Dask DataFrame, then it causes problems.

hayesgb replied:
@yaronha -- Are you comfortable with this?

hayesgb replied:
Per discusion, explicitly import dask.dataframe as dd, then load to dask if df_module is dd on line 198.

hayesgb commented on tests/datastore/test_base.py:
```diff
@@ -6,3 +7,10 @@ def test_http_fs_parquet_as_df():
         "https://s3.wasabisys.com/iguazio/data/market-palce/aggregate/metrics.pq"
     )
     data_item.as_df()
+
```
Based on the above, am I correct in assuming that storage_options are available as environmental variables?

Hedingber replied:
I don't think so, note this is not an s3 file path, but an http path, to a file in s3, it does not require any auth to get it.

hayesgb replied:
I've switched the test file for validating mlrun store objects will load into a Dask DataFrame a public container on Azure Open Datasets.  Is this acceptable?

hayesgb commented:
I would like to add at least one test for targets.py, but unsure how to implement as an artifact.

yaronha commented:
@Hedingber i talked to @hayesgb and the PR LGTM

Hedingber commented:
@hayesgb PR looks good.
I do think we better have unit test (`test_azure_blob.py` is integration) coverage for the code you wrote in `mlrun/datastore/targets.py`.
I believe you can easily achieve that by doing `log_dataset` (in the current `test_load_object_into_dask_dataframe` or in a new test in the same file) similar to what you did in the tests for azure, just note you'll need to give a big enough dataset, otherwise your code converts it to a pandas dataframe in the `DatasetArtifact` class initialization. For the artifact path you can use `str(tests.conftest.results)`

hayesgb commented:
@Hedingber -- I've added tests to validate the writing of a dask dataframe to a datastore, as well as tests to validate  the DatasetArtifacts methods.  Unfortunately, when I try to run tests that involve a large dataframe (>1GB) using `test-dockerized`, I'm getting a killed process error, so I added these to the integration suite.

Would appreciate any feedback you may have.

Hedingber commented:
@hayesgb Can you please align your branch to `development` ?
(As you can see your PR now shows above 2000 lines of code added, most of them are not really related to the PR)

Hedingber commented:
@hayesgb 
Regarding the test, to overcome what you described, I would suggest to do keep it as a unit test but to simply change the `max_ddf_memory` limit at the beginning of the test to something much smaller (so you'll easily exceed it) and then just don't forget to set it back at the end of the test

hayesgb commented:
@Hedingber -- The update failed here, but when I run `make test-dockerized` locally it is passing.  Can you provide guidance on why it may be failing here?  Logs suggest it's something in the mlrun scheduler.

Hedingber commented:
@hayesgb It's some transient CI failure we have, you can ignore it, triggered re-run and CI passed
Waiting for @yaronha to give final approval and I'll merge

hayesgb commented on mlrun/datastore/targets.py:
```diff
@@ -345,6 +345,13 @@ def write_dataframe(
+            try:
+                storage_options = self._get_store().get_storage_options()
+                df = df.repartition(partition_size="100MB")
+                df.to_parquet(self._target_path, storage_options=storage_options)
```
@yaronha -- Would you suggest moving this code into the `_write_dataframe()` method of the ParquetTarget class?  If that would work, then I could also add the same code (except df.to_csv()) to CSVTarget's _write_dataframe method.

Thoughts?

hayesgb replied:
As an aside -- I see that the rdd write method occurs in the BaseStoreTarget.  Am I correct in assuming that the difference between CSV and Parquet is handled in the kwargs?

hayesgb replied:
@yaronha -- what are your thoughts on this?  Separates CSV and Parquet, using a pattern similar to the rdd implementation.

Hedingber reviewed (approved):
Talked with @yaronha offline, changes looks good, approved

Hedingber commented:
@hayesgb PR merged, thanks!
